exporter.go

```diff
@@ -390,13 +437,54 @@ func (b *Exporter) handleEvent(event Event) {
 	}
 }
 
+// removeStaleMetrics removes label values set from metric with stale values
+func (b *Exporter) removeStaleMetrics() {
+	now := time.Now()
+	// delete timeseries with expired ttl
+	for metricName := range b.labelValues {
+		for hash, lvs := range b.labelValues[metricName] {
+			if lvs.ttl == 0 {
+				continue
+			}
+			if lvs.lastRegisteredAt.Add(lvs.ttl).Before(now) {
+				b.Counters.Delete(metricName, lvs.labels)
+				b.Gauges.Delete(metricName, lvs.labels)
+				b.Summaries.Delete(metricName, lvs.labels)
+				b.Histograms.Delete(metricName, lvs.labels)
+				delete(b.labelValues[metricName], hash)
+			}
+		}
+	}
+}
+
+// saveLabelValues stores label values set to labelValues and update lastRegisteredAt time and ttl value
+func (b *Exporter) saveLabelValues(metricName string, labels prometheus.Labels, ttl time.Duration) {
+	_, hasMetric := b.labelValues[metricName]
+	if !hasMetric {
+		b.labelValues[metricName] = make(map[uint64]*LabelValues)
+	}
+	hash := hashNameAndLabels(metricName, labels)
+	_, ok := b.labelValues[metricName][hash]
+	if !ok {
+		b.labelValues[metricName][hash] = &LabelValues{
+			labels: labels,
+			ttl:    ttl,
+		}
+	}
+	now := time.Now()
+	b.labelValues[metricName][hash].lastRegisteredAt = now
+	// Update ttl from mapping
+	b.labelValues[metricName][hash].ttl = ttl
```

Comment:
you keep separately checking the presence of something in a map (`b.labelValues[metricName]`, `b.labelValues[metricName][hash]`) but discarding the thing itself, then walking into the map again and again.

Would it be better here to keep the values around, and use them directly? Like

```
	metric, hasMetric := b.labelValues[metricName]
 	if !hasMetric {
                metric = make(map[uint64]*LabelValues)
 		b.labelValues[metricName] =metric
 	}
 	hash := hashNameAndLabels(metricName, labels)
 	metricLabelValues, hasMetricLabelValues := b.labelValues[metricName][hash]
        …
```

That way you can save a few map lookups, and I find it easier to follow which value is being used where.